File: Data Structures/Trees/AJs/binary-search-tree.py
# insert, dfs, bfs, find, contains
import sys
sys.path.append('C:/Users/User/Desktop/Algorithm Mastery/Data Structures/Stacks and Queues/AJs')
import my_queue
print(my_queue.Queue())

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
    
    def __repr__(self):
        # return self.__str__()
        return "Node()"

    def __str__(self):
        return str(self.__class__) + ": " + str(self.__dict__)

class BinarySearchTree:
    def __init__(self):
        self.root = None
    
    def __repr__(self):
        return "BinarySearchTree()"

    def __str__(self):
        return str(self.__class__) + ": " + str(self.__dict__)
    
    def insertRecursive(self, value):
        node = Node(value)
        if self.root is None:
            self.root = node
            return self
        def inserter(parent: Node, child: Node):
            if parent.value == child.value: return None
            if child.value > parent.value:
                if parent.right:
                    return inserter(parent.right, child)
                else:
                    parent.right = child
            if child.value < parent.value:
                if parent.left:
                    return inserter(parent.left, child)
                else:
                    parent.left = child
            return self
        
        return inserter(self.root, node)

    def insertIterative(self, value):
        node = Node(value)
        if self.root is None:
            self.root = node
            return self
        else:
            current = self.root
            while True:
                if current.value == node.value: return None
                current = current.right if (node.value > current.value) else current.left
                if current is None:
                    current = node
                    break
        return self

    def contains(self, value):
        if self.root is None:
            return False
        else:
            current = self.root
            while True:
                if current.value == value: return True
                current = current.right if (value > current.value) else current.left
                if current is None:
                    break
        return False

    def find(self, value):
        if self.root is None: return None
        else:
            current = self.root
            while True:
                if current.value == value: return current
                current = current.left if (value < current.value) else current.right
                if current == None:
                    return None

    #def bfs(self):
        # traverses the tree, touching each node only once
        # create a queue to store our elements to traverse (push + shift)
        # create a results array to store traversed nodes
        # put the first node in the queue
        # start a loop that continue as long as there are nodes left in the queue to traverse
        # pop the queue
        # add the element to the results array
        # if there is a left element, add it to the queue
        # same for the right
        # return the results after the loop
        #queue = 
        #return None

                    



bst = BinarySearchTree()
(bst.insertIterative(10))
(bst.insertRecursive(6))
bst.insertRecursive(15)
bst.insertRecursive(3)
bst.insertRecursive(20)
(bst.insertRecursive(8))
print(bst.find(100))

File: Data Structures/Trees/QJs/binary-search-tree.js
log = (val) => console.log(val)
class Node {
    constructor(value) {
        this.value = value
        this.left = null
        this.right = null
    }
}

class BinarySearchTree {
    constructor() {
        this.root = null
    }

    insertRecursive(value) {
        let node = new Node(value)

        const insertNode = (parent, child) => {
            if(child.value > parent.value) {
                if(!parent.right) {
                    parent.right = child
                } else {
                    return insertNode(parent.right, child)
                }            
            }
            if(child.value < parent.value) {
                if(!parent.left) {
                    parent.left = child
                } else {
                    return insertNode(parent.left, child)
                }
            } 
            
            if(parent.value == child.value) return undefined
            
            return this   
        }

        // starts at the root
        if(!this.root) {
            this.root = node
            return this
        } else {
            return insertNode(this.root, node)
        }       
    }

    insertIterative(value) {
        let node = new Node(value)
        // handle no root edge case
        // set current node
        // while current node.left }} .right is undefined
        // set node based on the conditions
        if(!this.root) {
            this.root = node
        } else {
            let currentNode = this.root

            while(true) {
                if(currentNode.value < value) {
                    if(!currentNode.right) {
                        currentNode.right = node
                        break
                    }
                    else currentNode = currentNode.right
                }

                if(currentNode.value > value) {
                    if(!currentNode.left) {
                        currentNode.left = node
                        break
                    } 
                    else currentNode = currentNode.left
                }

                if(currentNode.value == value) return undefined
            }
        }
        return this
    }

    find(value){
        if(this.root === null) return false;
        var current = this.root,
            found = false;
        while(current && !found){
            if(value < current.value){
                current = current.left;
            } else if(value > current.value){
                current = current.right;
            } else {
                found = true;
            }
        }
        if(!found) return undefined;
        return current;
    }
    
    contains(value){
        if(this.root === null) return false;
        var current = this.root,
            found = false;
        while(current && !found){
            if(value < current.value){
                current = current.left;
            } else if(value > current.value){
                current = current.right;
            } else {
                return true;
            }
        }
        return false;
    }


    bfs() {
        let queue = []
        let store = []
    
        queue.push(this.root)
        while(queue.length > 0) {
            let node = queue.shift()
            store.push(node)
            if(node.left) {
                queue.push(node.left)
            }
            if(node.right) {
                queue.push(node.right)
            }
        }
    
        return store.map(c => c.value)
    }

    dfsPreOrder() {
        let store = []
        let current = this.root

        // pick the root
        // push its value into the store
        // if it has a right, call seach
        // same with left
        const search = (node) => {
            store.push(node.value)
            if(node.left) search(node.left)
            if(node.right) search(node.right)
        }

        search(current)
        return store
    }

    dfsPostOrder() {
        let store = []
        let current = this.root

        // pick the root
        // push its value into the store
        // if it has a right, call seach
        // same with left
        const search = (node) => {
            if(node.left) search(node.left)
            if(node.right) search(node.right)
            store.push(node.value)
        }

        search(current)
        return store
    }

    dfsInOrder() {
        let store = []
        let current = this.root

        // pick the root
        // push its value into the store
        // if it has a right, call seach
        // same with left
        const search = (node) => {
            if(node.left) search(node.left)
            store.push(node.value)
            if(node.right) search(node.right)
        }

        search(current)
        return store
    }
}

let node = new Node(1)
let tree = new BinarySearchTree()
tree.insertRecursive(10)
tree.insertRecursive(6)
tree.insertRecursive(15)
tree.insertRecursive(3)
tree.insertRecursive(20)
tree.insertRecursive(8)
log(tree.bfs())
log(tree.dfsPreOrder())
log(tree.dfsPostOrder())
log(tree.dfsInOrder())
